src/Symfony/Component/HttpClient/Tests/NoPrivateNetworkHttpClientTest.php

```diff
@@ -0,0 +1,166 @@
+
+            // weird scenarios (e.g.: when trying to match ipv4 address on ipv6 subnet)
+            ['10.0.0.1', 'fc00::/7',   false],
+            ['fc00::1',  '10.0.0.0/8', false],
+        ];
+    }
+
+    /**
+     * @dataProvider getBlacklistData
+     */
+    public function testBlacklist(string $ipAddr, $subnets, bool $mustThrow): void
+    {
+        $content = 'foo';
+        $url = sprintf('http://%s/', 0 < substr_count($ipAddr, ':') ? sprintf('[%s]', $ipAddr) : $ipAddr);
+
+        if ($mustThrow) {
+            $this->expectException(TransportException::class);
+            $this->expectExceptionMessage(sprintf('IP "%s" is blacklisted for URL "%s".', $ipAddr, $url));
+        }
+
+        $previousHttpClient = $this->getHttpClientMock($url, $ipAddr, $content);
+        $client = new NoPrivateNetworkHttpClient($previousHttpClient, $subnets);
+        $response = $client->request('GET', $url);
+
+        if (!$mustThrow) {
+            $this->assertEquals($content, $response->getContent());
+            $this->assertEquals(200, $response->getStatusCode());
+        }
+    }
+
+    public function testCustomOnProgressCallback()
+    {
+        $ipAddr = '104.26.14.6';
+        $url = sprintf('http://%s/', $ipAddr);
+        $content = 'foo';
+
+        $executionCount = 0;
+        $customCallback = function (int $dlNow, int $dlSize, array $info) use (&$executionCount): void {
+            ++$executionCount;
+        };
+
+        $previousHttpClient = $this->getHttpClientMock($url, $ipAddr, $content);
+        $client = new NoPrivateNetworkHttpClient($previousHttpClient);
+        $response = $client->request('GET', $url, ['on_progress' => $customCallback]);
+
+        $this->assertEquals(1, $executionCount);
+        $this->assertEquals($content, $response->getContent());
+        $this->assertEquals(200, $response->getStatusCode());
+    }
+
+    public function testNonCallableOnProgressCallback()
+    {
+        $ipAddr = '104.26.14.6';
+        $url = sprintf('http://%s/', $ipAddr);
+        $content = 'bar';
+        $customCallback = sprintf('cb_%s', microtime(true));
+
+        $previousHttpClient = $this->getMockBuilder(HttpClientInterface::class)->getMock();
+        $client = new NoPrivateNetworkHttpClient($previousHttpClient);
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Option "on_progress" must be callable.');
+
+        $client->request('GET', $url, ['on_progress' => $customCallback]);
+    }
+
+    public function testConstructor()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('$subnets argument must be string, array or null.');
+
+        $previousHttpClient = $this->getMockBuilder(HttpClientInterface::class)->getMock();
+        new NoPrivateNetworkHttpClient($previousHttpClient, 3);
+    }
+
+    private function getHttpClientMock(string $url, string $ipAddr, string $content)
```

Comment:
Should we use the MockHttpClient instead?

Reply:
I'll update!

Reply:
I used MockHttpClient as much as I could.
testBlacklist and testCustomOnProgressCallback need to use my private getHttpClientMock because I make some assertions and MockHttpClient doesn't fill primary_ip on $info.